From 3dddea02c313fd1ac3ca5700f9b67149e19c40a1 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 15 Jun 2021 21:47:35 +0300
Subject: [PATCH] [SW-45909] fwupd: protect from "multi-map"

Do not allow to perform multiple map operations to prevent tampering
with security open/close operations on several memory areas at the same
time.

Change-Id: Ib1285b0f1fffa7a47294e9b313430974c077e945
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index a4844c7..f8631f2 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -492,11 +492,16 @@
 		break;
 	case HL_UPD_MAP_SRAM:
 		cmd_str = "map_sram";
-		module_fwupd_send_status(state, HL_UPD_STS_MAP_ACK, NULL);
-		break;
+		/* no break */
 	case HL_UPD_MAP_SPSRAM:
-		cmd_str = "map_spsram";
-		module_fwupd_send_status(state, HL_UPD_STS_MAP_ACK, NULL);
+		cmd_str = cmd_str ?: "map_spsram";
+		/* Protect from multi-MAP */
+		if (state->map_offset || state->map_size)
+			sts = HL_UPD_STS_VALID_ERR;
+		else
+			sts = HL_UPD_STS_MAP_ACK;
+
+		module_fwupd_send_status(state, sts, NULL);
 		break;
 	case HL_UPD_RUN_SRAM:
 		cmd_str = "run_sram";
